### PR TITLE
Fix Entropy reload

### LINKED ISSUE #251: Restarted sims deviate from uninterrupted.
I haven't used the restart feature much yet, but on a 512**3 hydro sim I ran a quick test earlier, restarting at z=2.25 and evolving to z=2 and below are the power spectra in the sim that had no restart vs the restarted one. The power on large scales agrees well, but in the non-linear regime there appears to be a non-negligible difference between the final outputs. 

mpgadget_restart

I guess this is could be some numerical rounding error that happens after the particle data is saved and reloaded, but not sure why this would result in consistently more non-linear power in the restarted sim, and to such a large degree.

### PR DESCRIPTION
**Pull request by sbird**
This fixes #251 by correctly initialising Entropy when restarting from a snapshot with DENSITY_INDEPENDENT_SPH off. We load it from the InternalEnergy array both with and without the compile flag to reduce regressions in future.

The SFR after a restart is still not exactly the same as not restarting (differs by about 0.1%) but it's close enough. I don't think we can get bit-exact equivalence because the act of forming stars changes the local density. Tested that it does fix #251 . This is an argument for removing the DENSITY_INDEPENDENT_SPH compile flag in the future.

Also a small refactoring of find_timestep that I did when I thought that was the problem.

### CODE REVIEW

File: libgadget/timestep.c [hunk 1]
```diff
@@ -122,31 +123,25 @@ set_global_time(double newtime) {
     IonizeParams();
 }
 
-/* This function assigns new timesteps to particles and PM */
+/* This function assigns new short-range timesteps to particles.
+ * If the argument restrictPMStep is true (on a PM step usually),
```
Comment (rainwoodman) [inline libgadget/timestep.c]: restrictPMStep?

**Comment by rainwoodman** [comment]
Why can't we save and load Entropy and InternalEnergy separately when there is no density_split_sph?

**Comment by sbird** [comment]
Because trying to load Entropy will break compatibility with existing snapshots. Unless there is a way to express "try loading Entropy if it is there, otherwise fall back to IE"?

File: libgadget/timestep.c (hunk 1 above)
Comment (sbird) [reply libgadget/timestep.c]: Sorry, rebase error.

**Comment by rainwoodman** [comment]
That's fair.

To address the data versioning problem in general, I think we shall add a Header field indicating the schema version -- the version number shall increase when the data schema is no longer compatible -- in which case we will simply die. Then in petaio.c we can hard code the version as a string, and to find the last version that can read the file we can look for that string in git history. 

This results a cleaner code than the 'if this then that' approach. Of course it means old data has to be manually migrated to the new software (or with separately provided python scripts) -- but it is easier to maintain than carrying the migration code with the mainline code base.

**Comment by sbird** [comment]
The problem here though is not that the schema changed. (Entropy and InternalEnergy do also contain the same information for density-entropy SPH, so arguably we would just be wasting disc space writing both). The problem is really that our snapshot schema depends on compile-time flags.

What do you think about trying to fix this? (in a separate PR)? The Black Hole arrays can be written out unconditionally, since if there are no BHs they will just be empty. Then the difficulty will be only EgyWtDensity for the DENSITY_INDEPENDENT_SPH, and we could try and find a solution for that. We already did something similar with the paramfiles for the same reason.

**Comment by sbird** [comment]
Also didn't mean to push that third commit yet.

**Comment by sbird** [comment]
Oh, I see. We would need the schema versioning to make the extra arrays unconditional.

I support adding a schema version, and using it to make our snapshot format better defined. But I don't like the idea of having to rewrite the data formats when the schema changes, since it would make bisecting bugs across a format change almost impossible. I think ideally we would have the snapshot advertise its format number and the code deals with that by sending it to the right reader. I'm not quite sure how to do that though.

**Comment by rainwoodman** [comment]
First, schema versioning is beyond this PR; I think this PR looks fine to me. It doesn't seem to break the pressure-entroppy restart.

You argue the disadvantage of lifting the schema migration out of the main application is losing the ability to bisect cross schema changes. However if we have the script to freely migrate snapshots to any schema version then we can add an extra conversion step in bisection. 

The conversion always happens, it's just maintaining the logic going from any version to the internal data schema in C (if blocks or vtable), or maintaining it in Python as a sequence of script files (data-migration-v1-v2, data-migration-v2-v3, and the reverses) -- I remember seeing similar structured tools in small web-frameworks. 

To minimize the need to add the data-migration step in bisection we'd prefer to change the schema when the code is stable; though this "principle" is unrealistic -- we likely only identify the need to change a schema when something is broken.

**Comment by sbird** [comment]
That's true. Thinking about it: in practice if we are just adding or moving arrays we can I think easily preserve some level of forward compatibility, where schema version N can be read by the code expecting N-1 (it just ignores the new arrays). And the python script wouldn't have to rewrite the whole snapshot on disc either. So that seems fine.

**Comment by sbird** [comment]
Shall I merge?

**Comment by rainwoodman** [comment]
Yes. Merge please.

**Comment by rainwoodman** [comment]
Switching from InternalEnergy for Entropy to Entropy for Entroy would bbreak forward compatibility.

**Comment by sbird** [comment]
Not if InternalEnergy was still there: the old code would just read InternalEnergy, ignoring the now-present Entropy array, and the new code would read the Entropy.

**Pull request merged**